### PR TITLE
Modinreach 347 : Record Contribution Errors Should not Cause Contribution Jobs to Halt

### PR DESCRIPTION
**Pull request by arin-epm**
[MODINREACH-347](https://issues.folio.org/browse/MODINREACH-347)

## Purpose
Record Contribution Errors Should not Cause Contribution Jobs to Halt

## Approach
1. New runInitialContribution method created.
2. ConcurrentMessageListenerContainer implemented to consume Kafka messages through custom listener.
3. 503 response from D2IR server handled by creating custom ServiceSuspendedException.
4. Retry Kafka messages implemented for handling 503 response from D2IR server.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/batch/KafkaItemReader.java [hunk 1]
```diff
@@ -57,11 +61,13 @@ public V read() {
     }
 
     if (consumerRecords.hasNext()) {
+      log.info("hasNext is true");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/KafkaItemReader.java]: This was for me debugging. You can remove.

File: src/main/java/org/folio/innreach/batch/KafkaItemReader.java [hunk 2]
```diff
@@ -57,11 +61,13 @@ public V read() {
...
       ConsumerRecord<K, V> rec = consumerRecords.next();
       recordProcessor.accept(rec);
       partitionOffsets.put(new TopicPartition(rec.topic(), rec.partition()), rec.offset());
       return rec.value();
     } else {
+      log.info("hasNext is false");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/KafkaItemReader.java]: Same

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 3]
```diff
@@ -5,23 +5,21 @@
...
 import static org.folio.innreach.batch.contribution.ContributionJobContextManager.endContributionJobContext;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Objects;
-import java.util.UUID;
-import java.util.concurrent.Future;
+import java.net.URI;
+import java.util.*;
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java:8]: Let's always try to avoid wildcard imports.

File: src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java [hunk 4]
```diff
@@ -69,7 +83,18 @@ private Deserializer<String> keyDeserializer() {
...
     return UUID.fromString(new String(rec.headers().lastHeader(ITERATION_JOB_ID_HEADER).value()));
+//    return UUID.fromString("fa34246b-86a6-4743-8d20-f368ef3242d7");
+  }
+
+  public KafkaUtil createKafkaConsumer() {
+
+    var consumerProperties = kafkaProperties.buildConsumerProperties();
+    consumerProperties.put(GROUP_ID_CONFIG, jobProperties.getReaderGroupId());
+
+    var topic = "folio.contrib.tester.innreach";
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java]: Think this topic is just for testing. The prod topic is different.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 5]
```diff
@@ -64,48 +68,66 @@ public class ContributionJobRunner {
 
   private static final List<UUID> runningInitialContributions = Collections.synchronizedList(new ArrayList<>());
 
-  @Async
-  public Future<Void> runInitialContributionAsync(UUID centralServerId, String tenantId, UUID contributionId, UUID iterationJobId) {
+  public void runInitialContributionAsync(UUID centralServerId, String tenantId, UUID contributionId, UUID iterationJobId) {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Should rename this method since it's no longer async.

File: src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java [hunk 6]
```diff
@@ -20,6 +21,11 @@ public Exception decode(String methodKey, Response response) {
       return new BadCredentialsException("Can't get InnReach access token. Key/Secret pair is not valid");
     }
 
+    if (HttpStatus.valueOf(response.status()).equals(HttpStatus.SERVICE_UNAVAILABLE)) {
+      log.debug("CentralServer is suspended for contribution");
+      return new ServiceSuspendedException("CentralServer is suspended for contribution. Try again later");
+    }
+
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java:28]: This is good, but I think this is only going to handle the exception if we are using version 3. We also need to check for the string that Justin provided in the response body for version 2.

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -5,11 +5,11 @@ spring:
     name: mod-inn-reach
   datasource:
     username: ${DB_USERNAME:postgres}
-    password: ${DB_PASSWORD:postgres}
+    password: ${DB_PASSWORD:root}
```
Comment (steveellis) [inline src/main/resources/application.yml]: This seems unrelated to this PR. Is it?

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -5,11 +5,11 @@ spring:
...
     url: jdbc:postgresql://${DB_HOST:localhost}:${DB_PORT:5432}/${DB_DATABASE:db}
   jpa:
     hibernate:
-      ddl-auto: none
+      ddl-auto: update
```
Comment (steveellis) [inline src/main/resources/application.yml]: Same as above. Is this related to this PR?

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -168,3 +168,6 @@ kafka:
       topic-pattern: (${environment}\.)(.*\.)inventory\.holdings-record
       concurrency: ${KAFKA_EVENTS_CONCURRENCY:2}
       data-type: org.folio.innreach.dto.Holding
+  backoff:
+    interval: 30000
+    max_failure: 5
```
Comment (steveellis) [inline src/main/resources/application.yml]: Thinking ahead to how we're going to want to do the same thing for the instances of consumers for _ongoing contributions_. If we set these globally in the app yaml are they going to apply to _every instance_ of our use of the  `ConcurrentMessageListenerContainer`?

Is the retry and max attempts different for initial contrib vs ongoing is what I'm asking. At first I thought yes they are, but now I'm not so sure. Perhaps they should be exactly alike.
 
But these values will not suffice for production I think. We need for it to retry for as long as 12 hours (I'm thinking), and as we have been saying the interval **must** be > than the default timeout from okhttp3 unless I'm missing something. To get the max_failure figure out how many times of the interval will get us to 12 hours.

File: src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java [hunk 10]
```diff
@@ -41,13 +44,24 @@ public class IterationEventReaderFactory {
...
+  @Qualifier("retryInterval")
+  private final Long interval;
+
+  @Qualifier("retryMaxAttempts")
+  private final Long maxAttempts;
+
+  @Qualifier("itemExceptionListener")
+  private final ContributionExceptionListener contributionExceptionListener;
+
   public KafkaItemReader<String, InstanceIterationEvent> createReader(String tenantId) {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java:54]: Let's remove this method entirely and its tests. Why keep it around?

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 11]
```diff
@@ -64,48 +68,66 @@ public class ContributionJobRunner {
...
-    runInitialContribution(context);
+    KafkaUtil tempKafkaConsumer = itemReaderFactory.createKafkaConsumer();
+
+    CustomAckMessageListener customAckMessageListener = new CustomAckMessageListener(new ContributionProcessor(this), context, new Statistics());
 
-    return new AsyncResult<>(null);
+    tempKafkaConsumer.startOrCreateConsumer(customAckMessageListener);
   }
 
+  @Deprecated
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Why deprecate? Just remove and if it is used in tests, remove those too.

File: src/main/java/org/folio/innreach/util/KafkaUtil.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+import java.net.SocketTimeoutException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.kafka.clients.consumer.ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG;
+
+@Log4j2
+@Setter
+@RequiredArgsConstructor
+public class KafkaUtil {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/util/KafkaUtil.java]: I know this name is taken from the sample code, but it doesn't really work well for our purposes. How about something like `InitialContributionJobConsumerContainer`? It's got stuff in it that is specific to the init contribution job like the error handlers so it probably won't be reused any other place.

So remove from util directory obviously and put in a better place.

File: src/main/java/org/folio/innreach/util/KafkaUtil.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  private final Map<String,Object> consumerProperties;
+  private final String topic;
+  private final Deserializer<String> keyDeserializer;
+  private final Deserializer<InstanceIterationEvent> valueDeserializer;
+
+  private final Long interval;
+
+  private final Long maxAttempts;
+
+  private final int THREAD_SIZE = 2;
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/util/KafkaUtil.java]: I think it's fine to hardcode this for our first test run with prod data. 2 should give us roughly twice as good perf and likely won't overtax the container.

File: src/main/java/org/folio/innreach/util/KafkaUtil.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      contributionExceptionListener.logWriteError(exception, record.value().getInstanceId());
+      log.info("Stopping consumer topic: {} after retry exhaustion", consumerRecord.topic());
+      stopConsumer(consumerRecord.topic());
+    }, fixedBackOff);
+    errorHandler.addRetryableExceptions(ServiceSuspendedException.class);
+    errorHandler.addRetryableExceptions(SocketTimeoutException.class);
+    return errorHandler;
+  }
+
+  public void startOrCreateConsumer(Object messageListner) {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/util/KafkaUtil.java]: For methods that attempt to do something (here start if it isn't already started) I like to prepend a `try` on the method name. So here I think this method should simply be `tryStart`. I think the word `consumer` is confusing/misleading here. We're basically starting the container which starts one or more consumers. If we rename the class as I recommend above, I think `tryStart` is clear.

File: src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java [hunk 15]
```diff
@@ -69,7 +83,18 @@ private Deserializer<String> keyDeserializer() {
   }
 
   private static UUID getJobId(ConsumerRecord<String, InstanceIterationEvent> rec) {
+    // TODO Comment this back in after testing
     return UUID.fromString(new String(rec.headers().lastHeader(ITERATION_JOB_ID_HEADER).value()));
+//    return UUID.fromString("fa34246b-86a6-4743-8d20-f368ef3242d7");
+  }
+
+  public KafkaUtil createKafkaConsumer() {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java]: Let's rename this method to `createInitialContributionConsumerContainer` since it can be more than one and I'm recommending a rename of `KafkaUtil` to be more in line with our purposes.

File: src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java (hunk 4 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java]: See how the old `createReader` constructs the topic:
```
var topic = String.format("%s.%s.%s", folioEnv.getEnvironment(), tenantId, jobProperties.getReaderTopic());
```

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 5 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: How about `startInitialContribution` since `runInitialContribution` is already taken.

File: src/main/java/org/folio/innreach/batch/contribution/service/CustomAckMessageListener.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+import lombok.AllArgsConstructor;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.folio.innreach.batch.contribution.ContributionJobContext;
+import org.folio.innreach.domain.dto.folio.inventorystorage.InstanceIterationEvent;
+import org.springframework.kafka.listener.AcknowledgingMessageListener;
+import org.springframework.kafka.support.Acknowledgment;
+
+@AllArgsConstructor
+public class CustomAckMessageListener implements AcknowledgingMessageListener<String, InstanceIterationEvent> {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/CustomAckMessageListener.java]: I think this is no longer an "acknowledging message listener" because it no longer acks (commits). I think two things should happen here:
* Go back to the sample code. It had an implementation that wasn't acknowledging and implement that instead here. Of course testing that it works.
* Rename this class to something more meaningful for our purposes like `InitialContributionMessageListener`.

File: src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java (hunk 6 above)
Comment (arin-epm) [reply src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java:28]: That part is handled in InnReachResponse object where we are getting 200 response while contributing & decontributing

File: src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+  private final Map<String,Object> consumerProperties;
+  private final String topic;
+  private final Deserializer<String> keyDeserializer;
+  private final Deserializer<InstanceIterationEvent> valueDeserializer;
+
+  private final Long interval;
+
+  private final Long maxAttempts;
+
+  private final int THREAD_SIZE = 2;
```
Comment (heorhiiii) [inline src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java]: _in general code could be prettier to increase its readability and maintainability_. E.g.: even within current example there are two omissions: 1) for such format it should be static final field; 2) for increasing meaning - it could be something like: "concurrent consumers amount" or "concurrent container size" as once noticed THREAD_SIZE constant devs who will maintain the codebase will likely imply Thread Pool, but it's just a property for concurrency of ConcurrentMessageListenerContainer.

File: src/main/java/org/folio/innreach/batch/KafkaItemReader.java [hunk 18]
```diff
@@ -46,6 +49,7 @@
   public void open() {
     if (kafkaConsumer == null) {
       kafkaConsumer = new KafkaConsumer<>(consumerProperties, keyDeserializer, valueDeserializer);
+     // new DefaultKafkaConsumerFactory<>(kafkaConsumer)
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/KafkaItemReader.java]: Should this line be removed?

File: src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+  private final Map<String,Object> consumerProperties;
+  private final String topic;
+  private final Deserializer<String> keyDeserializer;
+  private final Deserializer<InstanceIterationEvent> valueDeserializer;
+
+  private final Long interval;
+
+  private final Long maxAttempts;
+
+  private static final int CONCURRENT_CONTAINER_SIZE = 2;
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java]: Maybe just `CONCURENCY` would be better. Not sure if "size" really makes much sense here.

File: src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+    log.info("container is null");
+
+    ContainerProperties containerProps = new ContainerProperties(topic);
+
+    containerProps.setPollTimeout(100);
+    Boolean enableAutoCommit = (Boolean) consumerProperties.get(ENABLE_AUTO_COMMIT_CONFIG);
+//    if (!enableAutoCommit) {
+//      containerProps.setAckMode(ContainerProperties.AckMode.MANUAL_IMMEDIATE);
+//    }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java]: Should probably remove commented out code.

File: src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    container = new ConcurrentMessageListenerContainer<>(factory, containerProps);
+
+    container.setupMessageListener(messageListner);
+    container.setCommonErrorHandler(errorHandler());
+
+//    if (concurrency == 0) {
+//      container.setConcurrency(1);
+//    } else {
+//      container.setConcurrency(concurrency);
+//    }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java]: No longer needed so should remove.

File: src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    consumersMap.put(topic, container);
+
+    log.info("created and started kafka consumer for topic {}", topic);
+  }
+
+  public static void stopConsumer(final String topic) {
+    log.info("stopping consumer for topic {}", topic);
+    ConcurrentMessageListenerContainer<String, InstanceIterationEvent> container = consumersMap.get(topic);
+    container.stop();
+    log.info("consumer stopped!!");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/InitialContributionJobConsumerContainer.java]: Probably should add topic name to log message like above.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 23]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
       .centralServerId(centralServerId)
       .tenantId(tenantId)
       .build();
 
-    runInitialContribution(context);
+    totalRecords = numberOfRecords;
+
+    recordsProcessed.put(tenantId, 0);
+
+    InitialContributionJobConsumerContainer tempKafkaConsumer = itemReaderFactory.createInitialContributionConsumerContainer(tenantId);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: This variable name should probably be `container` or similar.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 24]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
-          }
-        }
-      });
+
+    var iterationJobId = context.getIterationJobId();
+
+    if (event == null) {
+      log.info("Exiting kafka message reader for contribution {} and job {}", contributionId, iterationJobId);
+      return;
+    }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java:128]: I don't think we want this block any more do we? Or maybe we want to keep the null check (so it doesn't blow up if it gets one), but change the log message to something more accurate to what would happen in this case.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 25]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
+  public void simulateContribution() {
+    makeSimulatedRequest();
+//    s.addRecordsTotal(1);
+  }
+
+  private void simulateContributionItems() {
+    // For now just make this one.
+    makeSimulatedRequest();
+//    s.addRecordsTotal(1);
+  }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: I think these two methods can be removed just to clean things up.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 26]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
+      String postBody = "somerandomstring"; // Should probably make this a random length.
+      URI testServer = URI.create("http://localhost:8080");
+      var res = testClient.makeTestRequest(testServer, postBody);
+      log.info("Response from test server: {}", res);
+    } catch (Exception e) {
+      log.warn("Error while simulating request: {} {}", e.getMessage(), e.getStackTrace());
+    } finally {
+      s.addRecordsProcessed(1);
+    }
+  }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: I think this method can be removed. It's not useful anymore is it? We have the one below which we can keep.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 27]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
+      }
+
+    } catch (ServiceSuspendedException ex) {
+      throw new ServiceSuspendedException(ex.getMessage());
+    } catch (Exception e) {
+      log.warn("Error while simulating request: {} {}", e.getMessage(), e.getStackTrace());
+    }
+  }
+
+  public Instance simulateLoadingInstance(Statistics s) {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Can this method be removed?

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 28]
```diff
@@ -64,57 +73,125 @@ public class ContributionJobRunner {
...
+      var res = testClient.makeTestRequest(testServer, postBody);
+      log.info("Response from test server: {}", res);
+    } catch (Exception e) {
+      log.warn("Error while simulating request: {} {}", e.getMessage(), e.getStackTrace());
+    } finally {
+      s.addRecordsProcessed(1);
+    }
+  }
+
+  private void makeSimulatedRequest() {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Let's keep this method around since it makes testing a lot easier.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionProcessor.java [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+public class ContributionProcessor implements IMessageProcessor{
+
+  private final ContributionJobRunner contributionJobRunner;
+
+  @Override
+  public void processMessage(String key, InstanceIterationEvent event, ContributionJobContext context, ContributionJobContext.Statistics statistics, String topic) {
+    System.out.println("Message is:"+ event.toString());
+    log.info("Processing initial contribution job {}", context);
+    contributionJobRunner.runInitialContribution(context, event, statistics, topic);
+//    contributionJobRunner.simulateContribution();
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionProcessor.java]: Maybe add a comment just above this commented out line that this is for testing using a local kafka and a local simulated d2ir server.

File: src/main/java/org/folio/innreach/batch/contribution/service/CustomAckMessageListener.java (hunk 16 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/batch/contribution/service/CustomAckMessageListener.java]: Don't think this is resolved Arin. Should this be an `AcknowledgingMessageListener` or just a `MessageListener` implementation. [This](https://github.com/bkatwal/kafka-util/blob/master/src/main/java/com/bkatwal/kafka/impl/CustomMessageListener.java#L10) is the sample code I'm referring to.

On line 26 you're no longer acknowledging. Because we have retry, we don't need to commit/ack, right? It's auto committing now, and retry is smart enough to do commit only when the retry succeeds. Have we verified that it properly retries and only commits on success?

File: src/main/java/org/folio/innreach/batch/contribution/service/InitialContributionMessageListener.java [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+
+import lombok.AllArgsConstructor;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.folio.innreach.batch.contribution.ContributionJobContext;
+import org.folio.innreach.domain.dto.folio.inventorystorage.InstanceIterationEvent;
+import org.springframework.kafka.listener.AcknowledgingMessageListener;
+import org.springframework.kafka.support.Acknowledgment;
+
+@AllArgsConstructor
+public class InitialContributionMessageListener implements AcknowledgingMessageListener<String, InstanceIterationEvent> {
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/InitialContributionMessageListener.java]: See my comment on previous versions. Should this be implementing this interface and not the `MessageListener` one?

File: src/main/java/org/folio/innreach/domain/service/impl/RecordContributionServiceImpl.java [hunk 31]
```diff
@@ -121,9 +132,19 @@ private InnReachResponse verifyBibContribution(UUID centralServerId, String bibI
...
-    Assert.isTrue(response.isOk(), "Unexpected items contribution response: " + response);
-    return response;
+    try {
+      var response = irContributionService.contributeBibItems(centralServerId, bibId, BibItemsInfo.of(bibItems));
+      if (!response.getErrors().isEmpty()) {
+        var error = response.getErrors().get(0).getReason();
+        if (error.contains("Contribution to d2irm is currently suspended")) {
+          throw new ServiceSuspendedException("Contribution to d2irm is currently suspended");
+        }
+      }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/RecordContributionServiceImpl.java]: This block is duplicated with the block above in `contributeBib`. Code duplication is always bad. Abstract this block into a method on the response object called `checkServiceSuspension` which throws our exception.

File: src/main/java/org/folio/innreach/domain/service/impl/RecordContributionServiceImpl.java [hunk 32]
```diff
@@ -109,9 +110,19 @@ private void logItemTransformationError(Item item, Exception e) {
...
+      Assert.isTrue(response.isOk(), "Unexpected contribution response: " + response);
+      return response;
+    } catch (ServiceSuspendedException ex) {
+      throw new ServiceSuspendedException(ex.getMessage());
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/RecordContributionServiceImpl.java]: So we throw an exception, catch the same exception, and re-throw it. What would be the purpose of doing that here? Throwing it once is all you need. There's no need for `try catch` here or in the other methods that do this.

File: src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java (hunk 6 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/external/client/feign/error/InnReachFeignErrorDecoder.java:28]: But why are handling this in two different places? I understand that one is a response code and one is a response body, but they are both part of the same response. Shouldn't error handling be consolidated in one place? See my comments below as well about code duplication.

File: src/main/java/org/folio/innreach/external/service/impl/InnReachContributionServiceImpl.java [hunk 33]
```diff
@@ -46,32 +47,54 @@ public InnReachResponse contributeBib(UUID centralServerId, String bibId, BibInf
...
+      if (!response.getErrors().isEmpty()) {
+        var error = response.getErrors().get(0).getReason();
+        if (error.contains("Contribution to d2irm is currently suspended")) {
+          throw new ServiceSuspendedException("Contribution to d2irm is currently suspended");
+        }
+      }
+      return response;
+    } catch (ServiceSuspendedException ex) {
+      throw new ServiceSuspendedException(ex.getMessage());
+    }
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/external/service/impl/InnReachContributionServiceImpl.java:99]: This is nearly the exact same code block as in `deContributeBib` above. Consolidate into a shared method.

File: src/main/resources/application.yml (hunk 9 above)
Comment (steveellis) [reply src/main/resources/application.yml]: How has this been resolved? I see that you have increased the interval and the max failure, but how do the new values you have chosen address my concerns?

File: src/main/resources/application.yml [hunk 34]
```diff
@@ -168,3 +168,6 @@ kafka:
       topic-pattern: (${environment}\.)(.*\.)inventory\.holdings-record
       concurrency: ${KAFKA_EVENTS_CONCURRENCY:2}
       data-type: org.folio.innreach.dto.Holding
+  backoff:
+    interval: 3600000
+    max_failure: 12
```
Comment (steveellis) [inline src/main/resources/application.yml]: These values are almost certainly not correct. Without these values being exactly right the entire solution that we have come up with won't work. Please read the documentation carefully for these values and come up with something that will work for our purposes.

As I have stated elsewhere: we want it to retry for at least 12 hours and retry needs to not conflict with okhttp3.

File: src/main/java/org/folio/innreach/batch/contribution/service/InitialContributionMessageListener.java [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+
+  @Override
+  public void onMessage(
+    ConsumerRecord<String, InstanceIterationEvent> consumerRecord) {
+
+    // process message
+    iMessageProcessor.processMessage(consumerRecord.key(), consumerRecord.value(), context, statistics, consumerRecord.topic());
+
+    // commit offset
+//    acknowledgment.acknowledge();
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/InitialContributionMessageListener.java]: Think this commented out bit can be removed now.

File: src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java [hunk 36]
```diff
@@ -72,4 +80,14 @@ private static UUID getJobId(ConsumerRecord<String, InstanceIterationEvent> rec)
...
   }
 
+  public InitialContributionJobConsumerContainer createInitialContributionConsumerContainer(String tenantId) {
+
+    var consumerProperties = kafkaProperties.buildConsumerProperties();
+    consumerProperties.put(GROUP_ID_CONFIG, jobProperties.getReaderGroupId());
+
+    var topic = String.format("%s.%s.%s",
+      folioEnv.getEnvironment(), tenantId, jobProperties.getReaderTopic());
+    return new InitialContributionJobConsumerContainer(consumerProperties,topic,keyDeserializer(),valueDeserializer(), retryConfig.getInterval(), retryConfig.getMaxAttempts(), contributionExceptionListener);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/IterationEventReaderFactory.java]: Need a space between the method parameters

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionProcessor.java [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+
+@Log4j2
+@AllArgsConstructor
+public class ContributionProcessor implements IMessageProcessor{
+
+  private final ContributionJobRunner contributionJobRunner;
+
+  @Override
+  public void processMessage(InstanceIterationEvent event, ContributionJobContext context, ContributionJobContext.Statistics statistics, String topic) {
+    System.out.println("Message is:"+ event.toString());
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionProcessor.java]: Don't forget to remove this

**Comment by Vignesh-kalyanasundaram** [comment]
this PR is not required. So, closing this PR.
